File: big_data_learning/spark-scala-java-learning/src/main/java/com/shawn/study/big/data/spark/java/core/rdd/basic/MapPartitionsWithIndexDemo.java
package com.shawn.study.big.data.spark.java.core.rdd.basic;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.apache.commons.lang.StringUtils;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class MapPartitionsWithIndexDemo {

  public static void main(String[] args) {
    SparkConf conf =
        new SparkConf().setAppName("map-partitions-with-index-demo").setMaster("local[*]");
    JavaSparkContext context = new JavaSparkContext(conf);
    JavaRDD<Integer> javaRDD = context.parallelize(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8), 3);
    List<Integer> list =
        javaRDD
            .mapPartitionsWithIndex(
                (index, iterator) -> {
                  if (index == 1) {
                    return iterator;
                  } else {
                    return Collections.emptyIterator();
                  }
                },
                false)
            .collect();
    System.out.println(StringUtils.join(list, ", "));
    context.close();
  }
}

File: big_data_learning/spark-scala-java-learning/src/main/java/com/shawn/study/big/data/spark/scala/rdd/api/transformation/MapPartitionsWithIndexDemo.scala
package com.shawn.study.big.data.spark.scala.rdd.api.transformation

import org.apache.spark.{SparkConf, SparkContext}

object MapPartitionsWithIndexDemo {

  def main(args: Array[String]): Unit = {
    val sparkConf = new SparkConf().setMaster("local[*]").setAppName("map-partitions-with-index-demo")
    val sc = new SparkContext(sparkConf)
    val rdd = sc.makeRDD(List(1,2,3,4), 2)
    // 【1，2】，【3，4】
    val mpiRDD = rdd.mapPartitionsWithIndex(
      (index, iter) => {
        if ( index == 0 ) {
          iter
        } else {
          Nil.iterator
        }
      }
    )
    mpiRDD.collect().foreach(println)
    sc.stop()
  }

}
